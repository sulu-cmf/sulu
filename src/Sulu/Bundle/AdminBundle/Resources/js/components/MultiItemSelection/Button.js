// @flow
import React from 'react';
import classNames from 'classnames';
import Icon from '../Icon';
import buttonStyles from './button.scss';

type Props = {
    icon: string,
    onClick: () => void,
    location: 'left' | 'right',
};

export default class Button extends React.PureComponent<Props> {
    handleClick = () => {
        this.props.onClick();
    };

    render() {
        const {
            icon,
            location,
        } = this.props;
        const buttonClass = classNames(
            buttonStyles.button,
            buttonStyles[location]
        );

        return (
            <button
                type="button"
                onClick={this.handleClick}
                className={buttonClass}
            >
                <Icon name={icon} />
            </button>
        );
    }
}
