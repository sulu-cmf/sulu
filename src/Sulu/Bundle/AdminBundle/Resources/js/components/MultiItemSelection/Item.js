// @flow
import React from 'react';
import type {Node} from 'react';
import {SortableHandle} from 'react-sortable-hoc';
import Icon from '../Icon';
import itemStyles from './item.scss';

const DRAG_ICON = 'su-more';
const REMOVE_ICON = 'su-trash-alt';

type Props = {
    id: string | number,
    index: number,
    children: Node,
    onRemove?: (id: string | number) => void,
};

export default class Item extends React.PureComponent<Props> {
    createDragHandle() {
        return SortableHandle(({children, className}) => (
            <span className={className}>{children}</span>
        ));
    }

    handleRemove = () => {
        if (this.props.onRemove) {
            this.props.onRemove(this.props.id);
        }
    };

    render() {
        const {
            index,
            onRemove,
            children,
        } = this.props;
        const DragHandle = this.createDragHandle();

        return (
            <div className={itemStyles.item}>
                <DragHandle className={itemStyles.dragHandle}>
                    <Icon name={DRAG_ICON} />
                    <span className={itemStyles.index}>{index}</span>
                </DragHandle>
                <div className={itemStyles.content}>
                    {children}
                </div>
                {onRemove &&
                    <button
                        type="button"
                        className={itemStyles.removeButton}
                        onClick={this.handleRemove}
                    >
                        <Icon name={REMOVE_ICON} />
                    </button>
                }
            </div>
        );
    }
}
