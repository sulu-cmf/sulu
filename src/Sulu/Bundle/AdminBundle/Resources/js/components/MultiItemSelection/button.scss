@import '../../containers/Application/colors.scss';

$buttonTextColor: $scorpion;
$buttonTextColorActive: $white;
$buttonBackgroundColor: $alto;
$buttonBackgroundColorActive: $shakespeare;
$buttonMargin: 2px;

.button {
    width: 40px;
    color: $buttonTextColor;
    background-color: $buttonBackgroundColor;
    border: 0;
    cursor: pointer;
    font-size: 14px;

    &:hover {
        color: $buttonTextColorActive;
        background-color: $buttonBackgroundColorActive;
    }

    &.left {
        margin-right: $buttonMargin;
    }

    &.right {
        margin-left: $buttonMargin;
    }
}
