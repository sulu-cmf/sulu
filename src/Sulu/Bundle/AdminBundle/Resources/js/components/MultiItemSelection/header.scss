@import '../../containers/Application/colors.scss';

$labelTextColor: $scorpion;
$contentBackgroundColor: $alto;

.header {
    display: flex;
    justify-content: space-between;
    height: 40px;
    overflow: hidden;
    border-radius: 3px 3px 0 0;

    &.empty-list {
        border-radius: 3px;
    }
}

.label {
    display: flex;
    flex-direction: column;
    justify-content: center;
    flex: 1;
    height: 100%;
    background-color: $contentBackgroundColor;
    font-size: 12px;
    color: $labelTextColor;
    padding: 0 20px;
    position: relative;
}

.loader {
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
}
