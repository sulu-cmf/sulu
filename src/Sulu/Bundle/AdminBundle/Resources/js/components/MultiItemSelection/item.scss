@import '../../containers/Application/colors.scss';

$itemBackgroundColor: $white;

.item {
    display: flex;
    justify-content: space-between;
    height: 40px;
    background-color: $itemBackgroundColor;
}

.content {
    display: flex;
    flex-direction: column;
    justify-content: center;
    flex: 1;
    height: 100%;
    margin: 0 20px 0 22px;
    font-size: 12px;
}

.remove-button {
    border: 0;
    padding: 0 10px;
    margin-right: 10px;
    cursor: pointer;
    background-color: transparent;
}

.drag-handle {
    display: flex;
    align-items: center;
    width: 40px;
    border: 0;
    padding-left: 20px;

    &.sortable {
        cursor: grab;
    }
}

.index {
    padding-left: 2px;
    display: inline-block;
    vertical-align: top;
    font-size: 12px;
}
