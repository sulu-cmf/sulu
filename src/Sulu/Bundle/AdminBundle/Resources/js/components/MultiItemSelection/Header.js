// @flow
import React from 'react';
import classNames from 'classnames';
import type {Button as ButtonConfig} from './types';
import Button from './Button';
import headerStyles from './header.scss';

type Props = {
    label?: string,
    emptyList: boolean,
    leftButton?: ButtonConfig,
    rightButton?: ButtonConfig,
};

export default class Header extends React.PureComponent<Props> {
    static defaultProps = {
        emptyList: true,
    };

    render() {
        const {
            label,
            emptyList,
            leftButton,
            rightButton,
        } = this.props;
        const headerClass = classNames(
            headerStyles.header,
            {
                [headerStyles.emptyList]: emptyList,
            }
        );

        return (
            <div className={headerClass}>
                {leftButton &&
                    <Button {...leftButton} location="left" />
                }
                <div className={headerStyles.label}>
                    {label}
                </div>
                {rightButton &&
                    <Button {...rightButton} location="right" />
                }
            </div>
        );
    }
}
