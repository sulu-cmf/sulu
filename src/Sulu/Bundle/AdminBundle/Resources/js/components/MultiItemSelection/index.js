// @flow
import MultiItemSelection from './MultiItemSelection';

export default MultiItemSelection;
