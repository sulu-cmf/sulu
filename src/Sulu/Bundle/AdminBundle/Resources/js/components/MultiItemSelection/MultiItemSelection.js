// @flow
import React from 'react';
import type {ChildrenArray, Element} from 'react';
import {SortableContainer, SortableElement} from 'react-sortable-hoc';
import type {Button} from './types';
import Header from './Header';
import Item from './Item';
import multiItemSelectionStyles from './multiItemSelection.scss';

type Props = {
    children: ChildrenArray<Element<typeof Item>>,
    /** The text inside the header bar of the `MultiItemSelection` */
    label?: string,
    /** Called when the remove button is clicked on an item */
    onItemRemove?: (itemid: string | number) => void,
    /** Called after a drag and drop action was executed */
    onItemsSorted?: (oldIndex: number, newIndex: number) => void,
    /** The config of the button placed left side of the header */
    leftButton?: Button,
    /** The config of the button placed right side of the header */
    rightButton?: Button,
};

export default class MultiItemSelection extends React.PureComponent<Props> {
    static Item = Item;

    createItem(originalItem: Element<typeof Item>) {
        return (
            <li className={multiItemSelectionStyles.listElement}>
                {
                    React.cloneElement(
                        originalItem,
                        {
                            ...originalItem.props,
                            onRemove: this.props.onItemRemove && this.handleItemRemove,
                        }
                    )
                }
            </li>
        );
    }

    createSortableList() {
        const SortableItem = this.createSortableItem();

        return SortableContainer(({children}) => (
            <ul className={multiItemSelectionStyles.list}>
                {React.Children.map(children, (item, index) => (
                    <SortableItem index={index}>
                        {item}
                    </SortableItem>
                ))}
            </ul>
        ));
    }

    createSortableItem() {
        return SortableElement(({children}) => {
            return this.createItem(children);
        });
    }

    handleItemRemove = (itemId: string | number) => {
        if (this.props.onItemRemove) {
            this.props.onItemRemove(itemId);
        }
    };

    handleItemsSorted = ({oldIndex, newIndex}: {oldIndex: number, newIndex: number}) => {
        const {onItemsSorted} = this.props;

        if (onItemsSorted) {
            onItemsSorted(oldIndex, newIndex);
        }
    };

    render() {
        const {
            label,
            children,
            leftButton,
            rightButton,
        } = this.props;
        const emptyList = !React.Children.count(children);
        const SortableList = this.createSortableList();

        return (
            <div>
                <Header
                    label={label}
                    emptyList={emptyList}
                    leftButton={leftButton}
                    rightButton={rightButton}
                />
                <SortableList
                    axis="y"
                    lockAxis="y"
                    useDragHandle={true}
                    onSortEnd={this.handleItemsSorted}
                    helperClass={multiItemSelectionStyles.dragging}
                >
                    {children}
                </SortableList>
            </div>
        );
    }
}
