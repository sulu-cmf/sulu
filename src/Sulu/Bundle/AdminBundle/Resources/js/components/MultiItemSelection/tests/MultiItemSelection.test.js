/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import {render, mount} from 'enzyme';
import MultiItemSelection from '../MultiItemSelection';

test('Render an empty MultiItemSelection', () => {
    expect(render(<MultiItemSelection label="I am empty" />)).toMatchSnapshot();
});

test('Render an MultiItemSelection with children', () => {
    expect(render(
        <MultiItemSelection label="I have children">
            <MultiItemSelection.Item
                id="1"
                index={1}
            >
                Child 1
            </MultiItemSelection.Item>
            <MultiItemSelection.Item
                id="2"
                index={2}
            >
                Child 2
            </MultiItemSelection.Item>
            <MultiItemSelection.Item
                id="3"
                index={3}
            >
                Child 3
            </MultiItemSelection.Item>
        </MultiItemSelection>
    )).toMatchSnapshot();
});

test('Clicking the left and right button inside the header should call the right handler', () => {
    const leftClickHandler = jest.fn();
    const rightClickHandler = jest.fn();
    const leftButtonConfig = {
        icon: 'plus',
        onClick: leftClickHandler,
    };
    const rightButtonConfig = {
        icon: 'gear',
        onClick: rightClickHandler,
    };
    const multiItemSelection = mount(
        <MultiItemSelection
            label="I have handler"
            leftButton={leftButtonConfig}
            rightButton={rightButtonConfig}
        >
            <MultiItemSelection.Item
                id="1"
                index={1}
            >
                Child 1
            </MultiItemSelection.Item>
            <MultiItemSelection.Item
                id="2"
                index={2}
            >
                Child 2
            </MultiItemSelection.Item>
            <MultiItemSelection.Item
                id="3"
                index={3}
            >
                Child 3
            </MultiItemSelection.Item>
        </MultiItemSelection>
    );

    expect(multiItemSelection.render()).toMatchSnapshot();

    multiItemSelection.find('.button.left').simulate('click');
    expect(leftClickHandler).toBeCalled();

    multiItemSelection.find('.button.right').simulate('click');
    expect(rightClickHandler).toBeCalled();
});

test('Clicking on the remove button inside an item should call the remove handler on the parent component', () => {
    const removeHandler = jest.fn();
    const clickedItemId = 1;
    const multiItemSelection = mount(
        <MultiItemSelection
            label="I have handler"
            onItemRemove={removeHandler}
        >
            <MultiItemSelection.Item
                id={clickedItemId}
                index={1}
            >
                Child 1
            </MultiItemSelection.Item>
            <MultiItemSelection.Item
                id="2"
                index={2}
            >
                Child 2
            </MultiItemSelection.Item>
            <MultiItemSelection.Item
                id="3"
                index={3}
            >
                Child 3
            </MultiItemSelection.Item>
        </MultiItemSelection>
    );

    multiItemSelection.find('.removeButton').at(0).simulate('click');
    expect(removeHandler).toHaveBeenCalledWith(clickedItemId);
});
