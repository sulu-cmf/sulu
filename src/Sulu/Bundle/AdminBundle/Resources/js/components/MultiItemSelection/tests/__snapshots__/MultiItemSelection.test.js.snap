// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clicking the left and right button inside the header should call the right handler 1`] = `
<div>
  <div
    class="header"
  >
    <button
      class="button left"
    >
      <span
        aria-label="plus"
        class="fa fa-plus"
      />
    </button>
    <div
      class="label"
    >
      I have handler
    </div>
    <button
      class="button right"
    >
      <span
        aria-label="gear"
        class="fa fa-gear"
      />
    </button>
  </div>
  <ul
    class="list"
  >
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle"
        >
          <span
            aria-label="ellipsis-v"
            class="fa fa-ellipsis-v"
          />
          <span
            class="index"
          >
            1
          </span>
        </span>
        <div
          class="content"
        >
          Child 1
        </div>
      </div>
    </li>
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle"
        >
          <span
            aria-label="ellipsis-v"
            class="fa fa-ellipsis-v"
          />
          <span
            class="index"
          >
            2
          </span>
        </span>
        <div
          class="content"
        >
          Child 2
        </div>
      </div>
    </li>
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle"
        >
          <span
            aria-label="ellipsis-v"
            class="fa fa-ellipsis-v"
          />
          <span
            class="index"
          >
            3
          </span>
        </span>
        <div
          class="content"
        >
          Child 3
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Render an MultiItemSelection with children 1`] = `
<div>
  <div
    class="header"
  >
    <div
      class="label"
    >
      I have children
    </div>
  </div>
  <ul
    class="list"
  >
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle"
        >
          <span
            aria-label="ellipsis-v"
            class="fa fa-ellipsis-v"
          />
          <span
            class="index"
          >
            1
          </span>
        </span>
        <div
          class="content"
        >
          Child 1
        </div>
      </div>
    </li>
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle"
        >
          <span
            aria-label="ellipsis-v"
            class="fa fa-ellipsis-v"
          />
          <span
            class="index"
          >
            2
          </span>
        </span>
        <div
          class="content"
        >
          Child 2
        </div>
      </div>
    </li>
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle"
        >
          <span
            aria-label="ellipsis-v"
            class="fa fa-ellipsis-v"
          />
          <span
            class="index"
          >
            3
          </span>
        </span>
        <div
          class="content"
        >
          Child 3
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Render an empty MultiItemSelection 1`] = `
<div>
  <div
    class="header emptyList"
  >
    <div
      class="label"
    >
      I am empty
    </div>
  </div>
  <ul
    class="list"
  />
</div>
`;
